File: 2022/day_11.js
const _ = require('lodash');
const utils = require('../utils');

function runOperation(monkey_num, worry) {
	var lcm = 3 * 13 * 19 * 17 * 5 * 7 * 11 * 2;
	switch(monkey_num){
		case 0:
			worry = (worry * 13) % lcm;
			monkey_throw = 1;
			if (worry % 3 == 0) {
				monkey_throw = 2;
			}
			break;
		case 1:
			worry = (worry + 2) % lcm;
			monkey_throw = 2;
			if (worry % 13 == 0) {
				monkey_throw = 7;
			}
			break;
		case 2:
			worry = (worry + 8) % lcm;
			monkey_throw = 7;
			if (worry % 19 == 0) {
				monkey_throw = 4;
			}
			break;
		case 3:
			worry = (worry + 1) % lcm;
			monkey_throw = 5;
			if (worry % 17 == 0) {
				monkey_throw = 6;
			}
			break;
		case 4:
			worry = (worry * 17) % lcm;
			monkey_throw = 3;
			if (worry % 5 == 0) {
				monkey_throw = 6;
			}
			break;
		case 5:
			worry = (worry + 3) % lcm;
			monkey_throw = 0;
			if (worry % 7 == 0) {
				monkey_throw = 1;
			}
			break;
		case 6:
			worry = (worry * worry) % lcm;
			monkey_throw = 0;
			if (worry % 11 == 0) {
				monkey_throw = 5;
			}
			break;
		case 7:
			worry = (worry + 6) % lcm;
			monkey_throw = 3;
			if (worry % 2 == 0) {
				monkey_throw = 4;
			}
			break;
	}
	
	return [worry, monkey_throw]
}

// game
var monkeys = [
	[54, 98, 50, 94, 69, 62, 53, 85],
	[71, 55, 82],
	[77, 73, 86, 72, 87],
	[97, 91],
	[78, 97, 51, 85, 66, 63, 62],
	[88],
	[87, 57, 63, 86, 87, 53],
	[73, 59, 82, 65]
];
var monkeyCounts = [0, 0, 0, 0, 0, 0, 0, 0];

// loop 20 rounds
for (var i = 1; i <= 10000; i++) {
	// loop each monkey
	for (var j = 0; j <= 7; j++) {
		// loop monkey items
		for (var k in monkeys[j]) {
			var result = runOperation(j, monkeys[j][k]);
			monkeyCounts[j]++;
			monkeys[result[1]].push(result[0]);
		}
		monkeys[j] = [];
	}
}
monkeyCounts.sort(function(a, b){return b-a});
console.log(monkeyCounts);
console.log(monkeyCounts[0] * monkeyCounts[1]);

File: 2022/day_11.py
from copy import deepcopy

file1 = open(r"C:\\AdventOfCode\2022\day_11.txt" ,"r")
bully = file1.readlines()

bullies = []
for i in bully:
    if "\n" in i:
        bullies.append(i[:-1])
    else:
        bullies.append(i)
        
bullies.append('')


decoded = []
for i in bullies:
    if "Monkey" in i:
        # Monkey num, starting items, worry operation, test for throw, to # if true, to # if false, count of items touched
        newmonk = [int(i.replace("Monkey ", "").replace(":", "")), None, None, None, None, None, 0]
    elif "Starting item" in i:
        working_items = i.replace("  Starting items: ", "").split(", ")
        working_items = [int(n) for n in working_items]
        newmonk[1] = working_items
    elif "Operation:" in i:
        op = i.replace("  Operation: new = ", "").split(" ")
        newmonk[2] = op
    elif "Test:" in i:
        divis = int(i.replace("  Test: divisible by ", ""))
        newmonk[3] = divis
    elif "true:" in i:
        newmonk[4] = int(i.replace("    If true: throw to monkey ", ""))
    elif "false:" in i:
        newmonk[5] = int(i.replace("    If false: throw to monkey ", ""))
    else:
        decoded.append(newmonk)

k = 3 * 13 * 19 * 17 * 5 * 7 * 11 * 2

def sanitytest_logical(item, test, divis, lc):
    
    #print([item, test, divis])

    
    if test[2]=="old":
        new = item * item
        
    elif test[1]=="*":
        new = item * int(test[2])
        
    elif test[1]=="+":
        new = item + int(test[2])
        
    new %= lc

    if new % divis==0:
        return [new, True]
    else:
        return [new, False]
        
    
def monkeybusiness_logical(script, rounds, LCM = None):
    
    script = deepcopy(script)
    
    for i in range(rounds):

        
        # do a turn for each monkey
        for j in range(len(script)):
            # skip if no items
            
            if script[j][1] == []:
                pass
            else:
                # for every item, k
                for k in range(len(script[j][1])):
                    
                    # add up each item as it is checked
                    script[j][6] += 1
                    # get new worry from k'th item of j'th row, use j'th row test, use j'th row divis
                    new_worry = sanitytest_logical(script[j][1][k], script[j][2], script[j][3], LCM)
                    
                    if new_worry[1]==True:
                        # go to the j'th true monkey's list, add lower_worry item
                        script[ script[j][4] ][1].append(new_worry[0])
                    else:
                        # go to the j'th false monkey's list, add lower_worry item
                        script[ script[j][5] ][1].append(new_worry[0])
                    
                # after a monkey goes through all their items, they should have none
                script[j][1] = []


    return script

testp2b = monkeybusiness_logical(decoded, rounds = 10000, LCM = k)
numTouches = []
for i in range(len(testp2b)):
    numTouches.append(testp2b[i][6])
numTouches.sort(reverse=True)
print(numTouches[0]*numTouches[1])